password-cli/app/Main.hs

```diff
@@ -0,0 +1,141 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Main (main) where
+
+import Control.Monad (join, void)
+import qualified Data.Password.Argon2 as Argon2
+import Data.Password.Bcrypt (PasswordCheck (..))
+import qualified Data.Password.Bcrypt as Bcrypt
+import qualified Data.Password.PBKDF2 as PBKDF2
+import qualified Data.Password.Scrypt as Scrypt
+import Data.Password.Types
+import qualified Data.Text.IO as T
+import Options.Applicative
+import System.Exit (exitFailure)
+import System.IO (stdin)
+
+main :: IO ()
+main = join $ execParser cliOpts
+
+cliOpts :: ParserInfo (IO ())
+cliOpts = info (commandsParser <**> helper) (fullDesc <> header "password CLI usage")
+  where
+    commandsParser :: Parser (IO ())
+    commandsParser =
+      subparser
+        ( command "argon2" (info commandArgon2 (progDesc "Argon2 operations"))
+            <> command "bcrypt" (info commandBcrypt (progDesc "Bcrypt operations"))
+            <> command "pbkdf2" (info commandPBKDF2 (progDesc "Pbkdf2 operations"))
+            <> command "scrypt" (info commandScrypt (progDesc "Scrypt operations"))
+            <> command "help" (info commandHelp (progDesc "Show command help"))
+        )
+
+commandHelp :: Parser (IO ())
+commandHelp =
+  go
+    <$> optional (argument str (metavar "COMMAND"))
+    <**> helper
+  where
+    go mCmd =
+      let args = maybe id (:) mCmd ["-h"]
+       in void $ handleParseResult $ execParserPure defaultPrefs cliOpts args
+
+commandArgon2 :: Parser (IO ())
+commandArgon2 =
```

Comment:
```hs
passwordCommand ::
    Parser (IO ()) ->
    (PasswordHash -> IO ()) ->
    Parser (IO ())
passwordCommand hashCommand checkCommand = 
  subparser
    ( command
        "hash"
        ( info
            hashCommand
            (progDesc "hash password (via STDIN)")
        )
        <> command
          "check"
          ( info
              (checkCommand <$> option str (metavar "HASH" <> long "hash") <**> helper)
              (progDesc "check hashed password (via STDIN)")
          )
    )
    <**> helper
```

And define the algorithm parsers as:
```hs
ommandBcrypt :: Parser (IO ())
commandBcrypt =
    passwordCommand
        (pure $ runHash Bcrypt.hashPassword)
        (runCheck . flip Bcrypt.checkPassword)
```

That way it's more DRY and when someone's going to add more options for specific algorithms, it's more obvious which parts are algorithm-specific, and what's common across all algorithms.

Reply:
Don't interpret this comment as a request for change, since @cdepillabout made a good point to rather parse to an `Option` type to then in the `main` function decide what to do depending on what.